substrate/client/consensus/beefy/src/communication/gossip.rs

```diff
@@ -283,9 +288,18 @@ where
 			let filter = self.gossip_filter.read();
 
 			match filter.consider_vote(round, set_id) {
-				Consider::RejectPast => return Action::Discard(cost::OUTDATED_MESSAGE),
 				Consider::RejectFuture => return Action::Discard(cost::FUTURE_MESSAGE),
 				Consider::RejectOutOfScope => return Action::Discard(cost::OUT_OF_SCOPE_MESSAGE),
+				Consider::RejectPast => {
+					// We know `vote` is for some past (finalized) block. Have fisherman check
+					// for equivocations. Best-effort, ignore errors such as state pruned.
+					let _ = self.fisherman.check_vote(vote);
+					// TODO: maybe raise cost reputation when seeing votes that are intentional
+					// spam: votes that trigger fisherman reports, but don't go through either
+					// because signer is/was not authority or similar reasons.
+					// The idea is to more quickly disconnect neighbors which are attempting DoS.
```

Comment:
IMO could worth a couple of words comment on why we are not calling `check_vote`/`check_proof` in `Consider::Accept` branch. Maybe that's obvious for those who's working on BEEFY, though :)

Reply:
> IMO could worth a couple of words comment on why we are not calling check_vote/check_proof in Consider::Accept branch. Maybe that's obvious for those who's working on BEEFY, though :)

Thanks for the good observation Slava :)
This was indeed not (fully) correctly implemented, see https://github.com/lederstrumpf/polkadot-sdk/compare/816b726c5f593a9d111d74ab0dc4757d7bd63a7f..3e9a4a74e6a for updated logic - the description of the logic is [here](https://github.com/lederstrumpf/polkadot-sdk/blob/3e9a4a74e6abe533c1db21d0183b06339b919cb8/substrate/client/consensus/beefy/src/communication/gossip.rs#L414-L424):
1. All votes and proofs (i.e. both for active and non-active rounds) are now checked for fork equivocations (since done across the board, I've moved this check to the calling [sc_consensus_beefy::communication::gossip::GossipValidator::validate(..)](https://github.com/lederstrumpf/polkadot-sdk/blob/3e9a4a74e6abe533c1db21d0183b06339b919cb8/substrate/client/consensus/beefy/src/communication/gossip.rs#L408-L458)).
2. As Adrian's pointed out, accepted votes are also checked later (namely in [`add_vote`](https://github.com/lederstrumpf/polkadot-sdk/blob/28af6f62d5a30e3de00ddb4cd89f42ad6e78de6a/substrate/client/consensus/beefy/src/round.rs#L122)), but for *vote* equivocations, so they should also be checked for fork equivocations (since an accepted vote may equivocate against a finalized payload without the voter having cast a prior vote for something else).
3. The reason we're checking proofs and votes for future non-active rounds (i.e. on blocks not finalized by grandpa yet) and flagging these as fork equivocations (despite them not equivocating against existing state) is that these should not have been voted on yet (hence, might be for/from an attack on the counterparty chain of a bridge)
4. Proofs for past and active rounds are for grandpa-finalized blocks, hence we should check whether they equivocate against our canonical chain.

> but we can only expect to see a `ForkEquivocation` for a block already finalized by us - aka on `RejectPast` - for `Consider::Accept` we know this is the first "valid" proof we're seeing (otherwise block would have been already finalized) - and for "invalid" proofs we don't check for equivocations as they are invalid anyway.

> but we can only expect to see a ForkEquivocation for a block already finalized by us ...

See `3.` above

> ... and for "invalid" proofs we don't check for equivocations as they are invalid anyway.

If you're using "invalid" as defined in `validate_finality_proof`'s [`verification block`](https://github.com/lederstrumpf/polkadot-sdk/blob/fa73cb5f0a4a5d69968a8562b38e5c3b4f80ab01/substrate/client/consensus/beefy/src/communication/gossip.rs#L372-L390):
Invalid proofs include those that pass [the trivial validity checks](https://github.com/lederstrumpf/polkadot-sdk/blob/fa73cb5f0a4a5d69968a8562b38e5c3b4f80ab01/substrate/client/consensus/beefy/src/justification.rs#L60-L65) but only have fewer than `2n/3+1` valid signatures. These are within our threat model: they are exactly the sorta proofs that cross-chain attackers would easily be able to craft (without violating our honest-majority assumption) when trying their luck to game BEEFY's subsampling protocol. So we should *primarily* be checking these ;)

> ... we know this is the first "valid" proof we're seeing (otherwise block would have been already finalized) ...

not entirely unambiguous to me how you're using "valid" or "finalized" here: We're only accepting proofs for active rounds, and all blocks in active rounds have already been (grandpa-)finalized - with "finalized", do you mean the block's within in the history of the [`best_beefy_block`](https://github.com/lederstrumpf/polkadot-sdk/blob/3e9a4a74e6abe533c1db21d0183b06339b919cb8/substrate/client/consensus/beefy/src/worker.rs#L92-L93)?

Reply:
> If you're using "invalid" as defined in validate_finality_proof's [verification block](https://github.com/lederstrumpf/polkadot-sdk/blob/fa73cb5f0a4a5d69968a8562b38e5c3b4f80ab01/substrate/client/consensus/beefy/src/communication/gossip.rs#L372-L390):
Invalid proofs include those that pass [the trivial validity checks](https://github.com/lederstrumpf/polkadot-sdk/blob/fa73cb5f0a4a5d69968a8562b38e5c3b4f80ab01/substrate/client/consensus/beefy/src/justification.rs#L60-L65) but only have fewer than 2n/3+1 valid signatures. These are within our threat model: they are exactly the sorta proofs that cross-chain attackers would easily be able to craft (without violating our honest-majority assumption) when trying their luck to game BEEFY's subsampling protocol. So we should primarily be checking these ;)

Yes, that is true, good catch!

> not entirely unambiguous to me how you're using "valid" or "finalized" here: We're only accepting proofs for active rounds, and all blocks in active rounds have already been (grandpa-)finalized - with "finalized", do you mean the block's within in the history of the [best_beefy_block](https://github.com/lederstrumpf/polkadot-sdk/blob/3e9a4a74e6abe533c1db21d0183b06339b919cb8/substrate/client/consensus/beefy/src/worker.rs#L92-L93)?

When considering a gossiped proof we get `Consider::Accept` **iff** the proof is _valid_ (disambiguation: contains more than two thirds validators signatures) **and** the block number it is for has not been already _finalized_ (disambiguation: not finalized by BEEFY. `block_number > best_beefy_block`). From this we can infer, that this is the first _valid_ BEEFY proof we're seeing for this block number.

Even so, it's still correct to check for equivocation since being the first doesn't guarantee it's also correct in terms of `payload`. So, yes, I support your change.

Reply:
But doing so, you also need to think about some newly opened scenarios:
- what about blocks between GRANDPA finalized and best imported? i.e. `best_grandpa < number < blockchain().info().best_block`

Thinking out loud:

Say our node is network lagging on gossip so it hasn't seen latest GRANDPA+BEEFY finality, and at some point a valid+correct BEEFY proof gets through to us for a block number where we locally haven't seen GRANDPA finality for yet. Meaning we might have multiple forks/hashes matching said block number - only one of which having correct `payload` => we might incorrectly treat this valid+correct proof as an equivocation.

But even if above happens, the ForkEquivationProof will be rejected on-chain when submitted right? because the payload from the proof must != from the on-chain payload at that block number - and on the "good" fork, that will not be the case - so we just slash on pruned forks where it doesn't matter.

^ Food for thought - maybe I missed something.

Reply:
> But even if above happens, the ForkEquivationProof will be rejected on-chain when submitted right? because the payload from the proof must != from the on-chain payload at that block number - and on the "good" fork, that will not be the case - so we just slash on pruned forks where it doesn't matter.

Your understanding is correct, see this note on `check_fork_equivocation_proof`: https://github.com/lederstrumpf/polkadot-sdk/blob/e7718fc8e75ed7133b3c06061374f32beefa7b8f/substrate/primitives/consensus/beefy/src/lib.rs#L492-L498

The most interesting place is indeed `best_grandpa < number < blockchain().info().best_block`:
If - like in your example - a node is lagging and sees BEEFY proofs on gossip before even seeing the block, then it will report this as an equivocation, but it will be ignored.
Likewise, without finality proofs, it will also get slashed on forks, which is immaterial.

https://github.com/lederstrumpf/polkadot-sdk/blob/e7718fc8e75ed7133b3c06061374f32beefa7b8f/substrate/client/consensus/beefy/src/communication/fisherman.rs#L300
I could tighten the above condition in the fisherman to report if `number < blockchain().info().finalized_number`, but since the runtime has no notion of finalization, I decided against this since it would just lead to reports that are either invalid (since a node is lagging on finality) or that the runtime would ignore until some other commitment is finalized:

If, as via the comment https://github.com/lederstrumpf/polkadot-sdk/blob/e7718fc8e75ed7133b3c06061374f32beefa7b8f/substrate/primitives/consensus/beefy/src/lib.rs#L492-L498, some nodes vote on commitments that have not been finalized yet, they will either
1. not be slashed if the commitment is ultimately finalized
2.  slashable once some other commitment is ultimately finalized for the same block height (since a valid equivocation proof can be generated at this stage)

which I deem covers the requirements.